app/gui/src/project-view/util/codemirror/persistence/index.ts

```diff
@@ -0,0 +1,133 @@
+import { useSyncLocalStorage, type SyncLocalStorageOptions } from '@/composables/syncLocalStorage'
+import {
+  scrollStatePlugin,
+  type ScrollState,
+  type ScrollStatePluginOptions,
+} from '@/util/codemirror/persistence/scroll'
+import type { Opt } from '@/util/data/opt'
+import type { ToValue } from '@/util/reactivity'
+import { type Extension } from '@codemirror/state'
+import { EditorView, ViewPlugin, type PluginValue } from '@codemirror/view'
+import { encoding } from 'lib0'
+import { nextTick, onBeforeUnmount, readonly, shallowRef, toValue, type ShallowRef } from 'vue'
+import { z } from 'zod'
+
+const editorSchema = z.object({
+  scroll: z.unknown(),
+})
+
+interface EditorState extends z.infer<typeof editorSchema> {
+  scroll?: ScrollState
+}
+
+interface EditorPersistencePluginOptions {
+  documentViewId: ToValue<Opt<string>>
+}
+/**
+ * Plugin that synchronizes editor state with local storage; the state is aggregated from any
+ * {@link PersistableStatePlugin}s (such as {@link scrollStatePlugin}) present in the view.
+ */
+class EditorPersistencePluginValue implements PluginValue {
+  private readonly documentViewId: ToValue<Opt<string>>
+  /**
+   * This will be changed to `true` when the `isConnected` state is first observed and the
+   * ready-callback scheduled.
+   */
+  private connected: boolean = false
+  /**
+   * This will be changed to `true` once the editor view is ready to be used. This prevents
+   * attempting to read or write DOM state, such as scroll position, before the editor DOM has been
+   * rendered.
+   */
+  private readonly ready: ShallowRef<boolean> = shallowRef(false)
+  /**
+   * Stores the last update attempted before the editor became ready, to be applied when the editor
+   * becomes ready.
+   */
+  private deferredUpdate: (() => void) | undefined = undefined
+  constructor(
+    private readonly view: EditorView,
+    { documentViewId }: EditorPersistencePluginOptions,
+  ) {
+    const { saveState } = useSyncLocalStorage(this.syncLocalStorageOptions())
+    onBeforeUnmount(saveState)
```

Comment:
The `onBeforeUnmount` save behavior might be moved into `useSyncLocalStorage`, seems like a sensible default.